server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/model/message/view/MessageFullViewFactory.java

```diff
@@ -79,6 +85,7 @@ public MessageFullView fromMetaDataWithContent(MetaDataWithContent message) thro
         Optional<String> textBody = computeTextBodyIfNeeded(messageContent, mainTextContent);
 
         Optional<Preview> preview = mainTextContent.map(Preview::compute);
+        storePreviewIfPossible(preview, message.getMessageId());
```

Comment:
Computing the preview is possibly an expensive operation (html parsing, mime parsing)...

Also I'm not confident having a write for every message read.

Can I suggest:

```
messagePreviewStore.retrieve(message.getMessageId())
    .switchIfEmpty(Mono.fromCallable(() -> mainTextContent.map(Preview::compute))
         .doOnNext(preview -> messagePreviewStore.store(message.getMessageId(), preview)
             .subscribe(Schedulers.boundedElastic())
             .blockOptional();
```